pi.py

```diff
@@ -1 +1,578 @@
+#
+#           ptop: Pressure below which sounding is ignored (hPa)
+#
+#           miss_handle: Flag that determines how missing (NaN) values are handled.
+#             If = 0 (BE02 default), NaN values in profile are ignored and PI is still calcuated
+#             If = 1 (pyPI default), given NaN values PI will be set to missing (with IFLAG=3)
+#             NOTE: If any missing values are between the lowest valid level and ptop
+#             then PI will automatically be set to missing (with IFLAG=3)
+#
+#
+#  OUTPUT:  CAPED (J/kg) is Convective Available Potential Energy of an air parcel
+#             consistent with its parcel and environmental properties.
+#
+#           TOB is the Temperature (K) at the level of neutral bouyancy 
+#             for the displaced air parcel
+#
+#           LNB is the pressure level of neutral bouyancy (hPa) for the 
+#             displaced air parcel
+#
+#           IFLAG is a flag where the value of 1 means OK; a value of 0
+#             indicates an improper sounding or parcel; a value of 2
+#             means that the routine failed to converge
+#
+
+    #
+    #   ***  Handle missing values   ***
+    #
+    
+    # find if any values are missing in the temperature or mixing ratio array
+    valid_i=~np.isnan(T)
+    first_valid=np.where(valid_i)[0][0]
+    # Are there missing values? If so, assess according to flag
+    if (np.sum(valid_i) != len(P)):
+        # if not allowed, set IFLAG=3 and return missing CAPE
+        if (miss_handle != 0):
+            CAPED=np.nan
+            TOB=np.nan
+            LNB=np.nan
+            IFLAG=3
+            # Return the unsuitable values
+            return(CAPED,TOB,LNB,IFLAG)
+        else:
+            # if allowed, but there are missing values between the lowest existing level
+            # and ptop, then set IFLAG=3 and return missing CAPE
+            if np.sum(np.isnan(T[first_valid:len(P)])>0):
+                CAPED=np.nan
+                TOB=np.nan
+                LNB=np.nan
+                IFLAG=3
+                # Return the unsuitable values
+                return(CAPED,TOB,LNB,IFLAG)
+            else:
+                first_lvl=first_valid
+    else:
+        first_lvl=0
+
+    # Populate new environmental profiles removing values above ptop and
+    # find new number, N, of profile levels with which to calculate CAPE
+    N=np.argmin(np.abs(P-ptop))
+    
+    P=P[first_lvl:N]
+    T=T[first_lvl:N]
+    R=R[first_lvl:N]
+    nlvl=len(P)
+    TVRDIF = np.zeros((nlvl,))
+    
+    #
+    #   ***  Run checks   ***
+    #
+
+    # CHECK: Is the input parcel suitable? If not, return missing CAPE
+    if ((RP < 1e-6) or (TP < 200)):
+        CAPED=0
+        TOB=np.nan
+        LNB=np.nan
+        IFLAG=0
+        # Return the unsuitable values
+        return(CAPED,TOB,LNB,IFLAG)
+    
+    #
+    #   ***  Define constants   ***
+    #
+
+    # Thermodynamic Constants
```

Comment:
I think these would be better defined in a separate file, constants.py or something of the sort. It's easier to keep all of these in one file.

Reply:
I love this idea, and will implement.